# Basic Pandas and I/O 

To practice using Pandas, we will explore the Atlas of Rural and Small Town America.  The USDA compiles county-level statistics from many different surveys to provide a comprehensive overview of different aspects of America.  This Atlas looks at broad categories of socioeconomic factors including the demographics of the population, economic data on employment, county properties on a rural-urban continuum, income data, and veteran status.  These categories are all stored as different sheets in an Excel file.  For this lab, we will only focus on the ``Jobs`` sheet.  For a full description of the available data, check out the webpage:

https://www.ers.usda.gov/data-products/atlas-of-rural-and-small-town-america/

Let's first import the Pandas library and load the data using the ``read_excel`` function.  Remember to specify the ``jobs`` sheet with the ``sheet_name`` argument.

In [1]:
import pandas as pd

df = pd.read_excel('../data/RuralAtlasData18.xlsx', sheet_name='Jobs')

Before doing any analysis, we first need to know what data fields we have and the type of data we will be working with.  Use ``.head()`` to get the first five rows of data.  If you just want to see a list of the columns are available, you can get the column names with ``.columns``.

In [2]:
# use .head on df
df.head()

FIPS State         County  UnempRate2017  UnempRate2016  UnempRate2015  \
0     0    US  United States            NaN            NaN            NaN   
1  1000    AL        Alabama            4.4            5.9            6.1   
2  1001    AL        Autauga            3.9            5.1            5.2   
3  1003    AL        Baldwin            4.0            5.4            5.5   
4  1005    AL        Barbour            5.9            8.4            8.9   

   UnempRate2014  UnempRate2010  UnempRate2007  PctEmpChange1017  ...  \
0            NaN            NaN            NaN               NaN  ...   
1            6.8           10.5            4.0          5.525260  ...   
2            5.8            8.9            3.3          6.303615  ...   
3            6.1           10.0            3.1         17.032748  ...   
4           10.5           12.3            6.3        -13.494810  ...   

   NumUnemployed2013  NumEmployed2013  NumCivLaborforce2013  \
0                NaN              NaN                   NaN   
1           156957.0        2017043.0             2174000.0   
2             1605.0          24205.0               25810.0   
3             5654.0          79626.0               85280.0   
4              931.0           8168.0                9099.0   

   NumUnemployed2007  NumEmployed2007  NumUnemployed2012  NumEmployed2012  \
0                NaN              NaN                NaN              NaN   
1            86485.0        2089127.0           173047.0        2003290.0   
2              806.0          23577.0             1779.0          23961.0   
3             2560.0          80099.0             6349.0          78065.0   
4              650.0           9684.0             1079.0           8283.0   

   UnempRate2012  NumCivLaborForce2012  NumUnemployed2014  
0            NaN                   NaN                NaN  
1            8.0             2176337.0           146531.0  
2            6.9               25740.0             1495.0  
3            7.5               84414.0             5300.0  
4           11.5                9362.0              932.0  

[5 rows x 62 columns]

This seems like a lot of data.  How much is it exactly?  We can get the size of the DataFrame in terms of rows and columns using ``.shape``.

In [3]:
print('DataFrame shape is:')
# use .shape  on sheet_data
df.shape

DataFrame shape is:


(3278, 62)

Pandas has automatically filled in its own index, but the unique identifier for our data is actually the Federal Information Processing Standards (FIPS) county code.  Let's set that as the index using ``.set_index()`` with the column we want specified as an argument.

In [4]:
df = df.set_index('FIPS')

Let's get into the summary statistics a bit.  It looks like there are 3278 counties in our dataset, but that definitely isn't right since there is a row for "United States" that appears when using ``.head()``.  Let's figure out how many counties there are in each state and see if there is anything else we don't want in the dataset.  The ``.value_counts()`` function is useful here and we can apply it to the ``State`` column.

In [5]:
df['State'].value_counts()

TX    255
GA    160
VA    135
KY    121
MO    116
KS    106
IL    103
NC    101
IA    100
TN     96
NE     94
IN     93
OH     89
MN     88
MI     84
MS     83
PR     79
OK     78
AR     76
WI     73
FL     68
AL     68
PA     68
SD     67
CO     65
LA     65
NY     63
CA     59
MT     57
WV     56
ND     54
SC     47
ID     45
WA     40
OR     37
AK     34
NM     34
UT     30
MD     25
WY     24
NJ     22
NV     18
ME     17
AZ     16
VT     15
MA     15
NH     11
CT      9
HI      6
RI      6
DE      4
DC      2
US      1
Name: State, dtype: int64

In addition to 'US', it looks like DC and Puerto Rico are in our data.  Filter them out since they aren't technically states.  "!=" means not equal.

In [6]:
df = df[(df['State'] != 'US') & (df['State'] != 'PR') & (df['State'] != 'DC')]

Print the value counts again to make sure they were taken out correctly before proceeding.

Now that we have a dataset that only contains records for counties that are in the 50 states, let's look at unemployment.  In particular we want to focus on the 2017 unemployment rate which can be accessed in the ``UnempRate2017`` column.  Let's pull it out into its own Series and assign it to ``UnempRate``.  If you recall the index will be maintained.

In [7]:
UnempRate = df['UnempRate2017']

Let's see if there are any missing values in our Series.  First, we can use ``.isna()`` to identify Null values with a Boolean flag (True or False) and then we can count them with ``.sum()``.  You can use these individually or chain them together.

In [8]:
print('Total Missing Values:')
UnempRate.isna().sum()

Total Missing Values:


6

Okay, it looks like we have 6.  To handle these Null values we could do something like filling it with the average value, but let's just drop them with ``.dropna()``.

In [9]:
UnempRate = UnempRate.dropna()

With this cleaned dataset, print the maximum unemployment rate, the minimum unemployment rate, and the average unemployment rate.

In [10]:
print('Maximum unemployment rate is:', UnempRate.max())
print('Minimum unemployment rate is:', UnempRate.min())
print('Average unemployment rate is:', UnempRate.mean())

Maximum unemployment rate is: 20.1
Minimum unemployment rate is: 1.6
Average unemployment rate is: 4.61166144200627


Let's actually see which counties have the maximum and minimum values.  We can do this by matching based off the column value.  For bonus points try to only print out the county and state.

In [11]:
print('The county with maximum unemployment rate is:')
print(df[df['UnempRate2017'] == UnempRate.max()][['County', 'State']])

The county with maximum unemployment rate is:
        County State
FIPS                
2158  Kusilvak    AK


In [12]:
print('The counties with minimum unemployment rate are:')
print(df[df['UnempRate2017'] == UnempRate.min()][['County', 'State']])

The counties with minimum unemployment rate are:
     County State
FIPS             
8009   Baca    CO
8125   Yuma    CO


Finally, let's sort the entire DataFrame by each county's unemployment rate (in ascending order).

In [13]:
df.sort_values(by=['UnempRate2017'])

State                           County  UnempRate2017  UnempRate2016  \
FIPS                                                                         
8125     CO                             Yuma            1.6            2.1   
8009     CO                             Baca            1.6            1.7   
8095     CO                         Phillips            1.7            2.0   
38091    ND                           Steele            1.7            2.1   
8017     CO                         Cheyenne            1.7            2.2   
38011    ND                           Bowman            1.7            2.2   
38023    ND                           Divide            1.7            2.6   
20071    KS                          Greeley            1.8            2.1   
8061     CO                            Kiowa            1.8            2.0   
8063     CO                       Kit Carson            1.8            2.1   
8057     CO                          Jackson            1.8            2.7   
19119    IA                             Lyon            1.8            2.0   
38021    ND                           Dickey            1.8            1.9   
38041    ND                        Hettinger            1.9            3.0   
48311    TX                         McMullen            1.9            2.1   
16065    ID                          Madison            1.9            2.4   
31029    NE                            Chase            1.9            2.1   
8121     CO                       Washington            1.9            2.1   
38073    ND                           Ransom            1.9            2.2   
38087    ND                            Slope            1.9            2.5   
38045    ND                          LaMoure            1.9            2.4   
8117     CO                           Summit            1.9            2.1   
31135    NE                          Perkins            2.0            2.1   
19169    IA                            Story            2.0            2.3   
19167    IA                            Sioux            2.0            2.3   
48205    TX                          Hartley            2.0            2.0   
38061    ND                        Mountrail            2.0            3.0   
8079     CO                          Mineral            2.0            2.2   
8051     CO                         Gunnison            2.0            2.3   
8047     CO                           Gilpin            2.0            2.4   
...     ...                              ...            ...            ...   
26119    MI                      Montmorency           10.8           10.2   
48489    TX                          Willacy           11.0           12.2   
53019    WA                            Ferry           11.0           10.2   
48507    TX                           Zavala           11.1           13.9   
22035    LA                     East Carroll           11.1           11.5   
1131     AL                           Wilcox           11.4           14.4   
2198     AK            Prince of Wales-Hyder           11.4           12.3   
28055    MS                        Issaquena           11.6           14.0   
48427    TX                            Starr           11.7           13.3   
46102    SD                    Oglala Lakota           11.8           10.8   
2164     AK               Lake and Peninsula           11.9           12.6   
2180     AK                             Nome           12.6           12.8   
2105     AK                    Hoonah-Angoon           12.6           11.9   
30003    MT                         Big Horn           13.4            7.3   
2050     AK                           Bethel           13.8           14.3   
35029    NM                             Luna           14.1           14.5   
6011     CA                           Colusa           14.3           15.6   
28063    MS                        Jefferson           14.7           14.9   
21153    KY                 